### PR TITLE
`azurerm_consumption_budget_*` refactor into base and scoped resources

### PR DESCRIPTION
**Pull request by stephybun**
The consumption resources were becoming a bit unwieldy hence the need for a refactor - this should also make it easier to swap over in the future.

The main changes contained in this PR are:
- Refactored all the resources into a base + a resource for each scope
- Moved all the flatten/expand functions into base and turned into private functions
- Added backwards compatibility for `azurerm_consumption_budget_subscription_id` since the validation on it is enforcing an incorrect and inconsistent behaviour with respect to the other scoped resources

The tests are practically unchanged.

image

### CODE REVIEW

File: internal/services/consumption/consumption_budget_base.go [hunk 1]
```diff
@@ -0,0 +1,841 @@
...
+						Type:         pluginsdk.TypeString,
+						Optional:     true,
+						Computed:     true,
+						ValidateFunc: validation.IsRFC3339Time,
+					},
+				},
+			},
+		},
+	}
+	for k, v := range fields {
```
Comment (tombuildsstuff) [inline internal/services/consumption/consumption_budget_base.go:315]: it'd be worth adding a comment about the override here

File: internal/services/consumption/consumption_budget_base.go [hunk 2]
```diff
@@ -0,0 +1,841 @@
...
+
+func (br consumptionBudgetBaseResource) attributes() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{}
+}
+
+// TODO remove in 3.0
+func parseScope(scope string) (string, error) {
+	// The validation behaviour of subscription_id isn't correct, it should only accept
+	// the resource ID format for the subscription. This ensures backward compatibility
+	// for the time being but should be removed in 3.0
```
Comment (tombuildsstuff) [inline internal/services/consumption/consumption_budget_base.go:330]: presumably we'll also need a state migration for this?

File: internal/services/consumption/consumption_budget_base.go [hunk 3]
```diff
@@ -0,0 +1,841 @@
...
+	timePeriod := make([]interface{}, 0)
+
+	if input == nil {
+		return timePeriod
+	}
+
+	timePeriodBlock := make(map[string]interface{})
+
+	timePeriodBlock["start_date"] = input.StartDate.String()
+	timePeriodBlock["end_date"] = input.EndDate.String()
```
Comment (tombuildsstuff) [inline internal/services/consumption/consumption_budget_base.go:565]: aren't these nilable?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @stephybun 

Thanks for this PR - I've taken a look through and left a few comments inline but if we can fix those up then this otherwise LGTM 👍 

Thanks!

**Comment by stephybun** [comment]
Thanks for the review @tombuildsstuff. Should be ready for another look over :)

**Review by katbyte — APPROVED** [review]
Thanks @stephybun - LGTM 🏗️

**Pull request merged**